Plot overlay images

TomSlater opened the pull request:
### Description of the change
This PR implements a function to provide overlayed images of different colours (for example, for elemental maps). It aims to resolve issue #1812 .

Further improvements could be made by using ipywidgets to select colours and alpha values but this just adds initial functionality.

### Progress of the PR
- [X] Change implemented (can be split into several points),
- [X] update docstring (if appropriate),
- [X] update user guide (if appropriate),
- [ ] add entry to `CHANGES.rst` (if appropriate),
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np
from urllib.request import urlretrieve, urlopen
from zipfile import ZipFile
files = urlretrieve("https://www.dropbox.com/s/ecdlgwxjq04m5mx/HyperSpy_demos_EDS_TEM_files.zip?raw=1./HyperSpy_demos_EDX_TEM_files.zip")
with ZipFile("HyperSpy_demos_EDX_TEM_files.zip") as z:
    z.extractall()
si_EDS = hs.load("core_shell.hdf5")
im = si_EDS.get_lines_intensity()
hs.plot.plot_overlay_images(im, colors=['r','b'], alphas=[1.0,1.0], legend_list=['Fe','Pt'], scalebar=True, axes_decor='off')
```

TomSlater committed:
Add plot_overlay_images function

TomSlater committed:
Add function to utils/plot import

TomSlater committed:
Add documentation

TomSlater committed:
Added basic test

TomSlater commented:
I've added a basic test for plotting but I can't seem to get pytest to create the plot when running with --mpl-generate-path (it is creating others but not the one I've added). Perhaps someone could help me with this?

ericpre commented:
I am wondering if it would make sense to incorporate it in `plot_images` instead of having a separate function?

> I've added a basic test for plotting but I can't seem to get pytest to create the plot when running with --mpl-generate-path (it is creating others but not the one I've added). Perhaps someone could help me with this?

Please check that your version of matplotlib and freetype match the one defined in `conda_environment_dev.yml` (https://hyperspy.readthedocs.io/en/latest/dev_guide/testing.html#plot-testing)

thomasaarholt commented:
> Please check that your version of matplotlib and freetype match the one defined in `conda_environment_dev.yml` (https://hyperspy.readthedocs.io/en/latest/dev_guide/testing.html#plot-testing)

It can be useful to create a testing evironment in the same way that @ericpre's nice github actions do it:

From the base hyperspy path, create and update a conda environment like this:

```bash
conda env create -f conda_environment.yml # install hyperspy dependencies
conda env update -f conda_environment_dev.yml # install hyperspy testing dependencies
conda activate test_env # test_env is the name set in the yml file
pip install -e . # install hyperspy as a develop install
```

@ericpre would it be a good idea to add the above to the docs? As an alternative to the first code block in [the "Running Tests" section](https://hyperspy.readthedocs.io/en/latest/dev_guide/testing.html#running-tests)?

TomSlater commented:
> I am wondering if it would make sense to incorporate it in `plot_images` instead of having a separate function?

I'm happy to incorporate it in to `plot_images` with an `overlay = True` argument or something similar. I don't have a preference really, I'd just followed Francisco's notes in #1812 when porting from my own function. Which do people think makes the most sense?

I'll have another go setting up a testing environment.

ericpre commented:
> > Please check that your version of matplotlib and freetype match the one defined in `conda_environment_dev.yml` (https://hyperspy.readthedocs.io/en/latest/dev_guide/testing.html#plot-testing)
> 
> It can be useful to create a testing evironment in the same way that @ericpre's nice github actions do it:
> 
> From the base hyperspy path, create and update a conda environment like this:
> 
> ```shell
> conda env create -f conda_environment.yml # install hyperspy dependencies
> conda env update -f conda_environment_dev.yml # install hyperspy testing dependencies
> conda activate test_env # test_env is the name set in the yml file
> pip install -e . # install hyperspy as a develop install
> ```
> 
> @ericpre would it be a good idea to add the above to the docs? As an alternative to the first code block in [the "Running Tests" section](https://hyperspy.readthedocs.io/en/latest/dev_guide/testing.html#running-tests)?

Maybe a link to https://hyperspy.readthedocs.io/en/latest/user_guide/install.html#installation-in-a-anaconda-minconda-distribution?

TomSlater committed:
Updated to return figure for tests

ericpre commented:
Here is an example:

![image](https://user-images.githubusercontent.com/11851990/102275569-7be2e180-3f1d-11eb-8a5a-b3a3163155df.png)

A few suggestions:
- since there is no need for a colorbar, we can remove the white padding around the image, when `axes_decor=off`. This is convenient when saving the figure.
- add/remove a map when clicking on the legend, as it is currently done with `plot_spectra`? This should be possible by putting the for loop in a callback function and adapting the alpha accordingly. This would be used for a future widget adjusting the alpha.
- the fact that there are quite a bit of code which is the same or very similar to the `plot_images` function convinced me that it should be in the same function. The idea would be to switch overlay on/off simply by changing one parameter.
- the `legend_title` should be the same as the `label` parameter of `plot_images` and there should be possible to find the label automatically from the title. `plot_images` should be already doing some of it.

TomSlater committed:
Updated to use plot_images

Encompassed the overlay utility in to the plot_images function and removed plot_overlay_images.

Added the ability to click on the legend to turn off/on different images.

Updated the cmap so the background is always black.

Updated to take the legend names from the function's label input.

TomSlater commented:
> A few suggestions:
> 
> * since there is no need for a colorbar, we can remove the white padding around the image, when `axes_decor=off`. This is convenient when saving the figure.
> * add/remove a map when clicking on the legend, as it is currently done with `plot_spectra`? This should be possible by putting the for loop in a callback function and adapting the alpha accordingly. This would be used for a future widget adjusting the alpha.
> * the fact that there are quite a bit of code which is the same or very similar to the `plot_images` function convinced me that it should be in the same function. The idea would be to switch overlay on/off simply by changing one parameter.
> * the `legend_title` should be the same as the `label` parameter of `plot_images` and there should be possible to find the label automatically from the title. `plot_images` should be already doing some of it.

I think I've covered almost all of these now, thanks for suggestions @ericpre ! It's incorporated in to `plot_images`, uses `label` for the legend and can now add/remove maps by clicking on the legend. You can get an overlay by simply setting  `overlay=True` in the arguments of `plot_images`.

I'm not quite sure how to get rid of whitespace around the image, I think it may work in some cases but not all now. If anyone has any suggestions that would be very helpful.

ericpre commented:
One way to do it would be:

```python
import numpy as np
import matplotlib.pyplot as plt

pixel_number = 500
data = np.arange(pixel_number**2).reshape(pixel_number, pixel_number)
shape = data.shape

dpi = 100
sampling_factor = 10 # adjust the width of single pixel
fig = plt.figure(figsize=[sampling_factor*v/dpi for v in shape], dpi=dpi)

ax = fig.add_axes([0, 0, 1, 1])
ax.axis('off')
ax.imshow(data, cmap='gray')

plt.savefig('test.png')
```

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -488,8 +490,15 @@ def plot_images(images,
+                overlay=False,
+                overlay_title=None,
+                colors=[],
+                alphas=[],
```
Set defaults value to `[1.0, 1.0, 1.0]`?

TomSlater replied:
I could set to a single value of 1.0? I could also take a single float value as well as a list, if the user wants all alphas to be the same.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -488,8 +490,15 @@ def plot_images(images,
                 vmax=None,
+                overlay=False,
+                overlay_title=None,
+                colors=[],
```
set defaults colors to `['r', 'b', 'g']`?

TomSlater replied:
I was trying to have a default that allowed any number of images (same with alphas). Maybe this should be set to list(BASE_COLORS) as default? Is that descriptive enough for the user?

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -601,6 +610,29 @@ def plot_images(images,
+    alphas : list of floats, optional
+        `alphas` should be a list of floats corresponding to the alpha
+		value of each color.
+    legend: {None, list of str, 'auto'}, optional
```
This is doing more or less the same as the `label` argument?

TomSlater replied:
True, I'll remove this argument.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+                raise ValueError("Images are not the same scale and so should"
+                                 "not be overlayed.")
+        
+        original_cmap = plt.rcParams['image.cmap']
+        plt.rcParams['image.cmap'] = 'gray'
```
Why is it necessary?

TomSlater replied:
This was to always set the background to black (rather than purple if the default cmap is viridis). I could add an argument to set cmap in case the user wants a different background colour. Or I could just remove and add a note on changing the default cmap if you want a different background colour.

ericpre replied:
I am not convinced this is worth making the background color adjustable? See other comment below to avoid changing the `rcParams`.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+            for i in range(len(images)):
+                alphas.append(1.0)
+    
+        ax.imshow(np.zeros_like(images[0].data))
```
Why is it necessary?

TomSlater replied:
I think this is needed to set the background colour (due to the transparency of the other colormaps).

ericpre replied:
Ok, would it be better to use `ax.imshow(np.zeros_like(images[0].data), cmap='gray')` and avoid setting the matplotlib `rcParams`?

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+        ax = f.add_subplot(1, 1, 1)
+        patches = []
+        
+        #If no colors are selected use BASE_COLORS
+        if colors == []:
+            for i in range(len(images)):
+                colors.append(list(BASE_COLORS)[i])
```
Define it in the definition of the function: less code for us and more explicit for the user.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+        original_cmap = plt.rcParams['image.cmap']
+        plt.rcParams['image.cmap'] = 'gray'
+        import matplotlib.patches as mpatches
+        ax = f.add_subplot(1, 1, 1)
```
Use `ax = f.add_axes([0, 0, 1, 1])` with the figure size set the correct aspect ratio as mentioned in https://github.com/hyperspy/hyperspy/pull/2599#issuecomment-746963107

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+            if legend_picking is True:
+                animate_legend(fig=f, ax=ax, plot_type='images')
+        
+        if axes_decor == 'off':
+            ax.margins=(0,0)
```
Remove this once the padding is fixed?

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+            my_cmap = cmap(np.arange(cmap.N))
+            my_cmap[:,-1] = np.linspace(0.3, 1, cmap.N)
+            my_cmap = ListedColormap(my_cmap)
+            ax.imshow(im.data, vmin=im.data.min(), vmax=im.data.max(),
```
The `vmin` and `vmax` defined in the function should be used here.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -878,135 +911,206 @@ def check_list_length(arg, arg_name):
+        for i, im in enumerate(images):
+            cols = ['k', colors[i]]
+            cmap=LinearSegmentedColormap.from_list('cmap'+str(i), cols)
+            my_cmap = cmap(np.arange(cmap.N))
+            my_cmap[:,-1] = np.linspace(0.3, 1, cmap.N)
+            my_cmap = ListedColormap(my_cmap)
```
Is this necessary? The following seems to be doing what we need?
```python
import numpy as np
from matplotlib.colors import LinearSegmentedColormap
import matplotlib.pyplot as plt

# cmap = LinearSegmentedColormap.from_list("", ["black", "red"])
cmap = LinearSegmentedColormap.from_list("", ["white", "red"])

plt.figure()
plt.imshow(np.arange(100).reshape((10, 10)), cmap=cmap)
plt.colorbar()
```

TomSlater replied:
My code adds a transparency to the colormaps so that you can see the different images when overlayed. If you use the code as you've written above, you will only see one image at a time (although you could play with the alphas to show others).

TomSlater replied:
Sorry, I should have said it adds a transparency gradient.

ericpre replied:
Ah ok, I missed that important bit! Currently, it looks a bit cumbersome the way the cmap is defined, what's about the following:
```python
import numpy as np
from matplotlib.colors import LinearSegmentedColormap, to_rgba
import matplotlib.pyplot as plt

def transparent_single_color_cmap(color):
    """ Return a single color matplotlib cmap with the transparency increasing
    linearly from 0 to 1."""
    return LinearSegmentedColormap.from_list("", [to_rgba(color, 0), to_rgba(color, 1)])

plt.figure()
data = np.arange(100).reshape((10, 10))
plt.imshow(data, cmap=transparent_single_color_cmap('r'))
plt.imshow(-data, cmap=transparent_single_color_cmap('b'))
```

ericpre reviewed (changes requested):
This is in good shape, I made suggestions to simplify the code.

TomSlater committed:
Update following suggestions

Added vmin and vmax use.

Removed use of legend argument.

Changed figure generation so will have no whitespace.

Changed default alphas and colors arguments.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -848,7 +844,13 @@ def __check_single_colorbar(cbar):
+        if overlay:
+            shape = images[0].data.shape
+            dpi = 100
+            sampling_factor = 10
```
I think it would be convenient to have the `sampling_factor` as a parameter, so that user can adjust the number of pixel of the image and save it from a script very easily. Not sure that `sampling_factor` is the most appropriate name?!

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -925,32 +927,45 @@ def check_list_length(arg, arg_name):
+            _vmax = vmax[idx] if isinstance(vmax, (tuple, list)) else vmax
+            _vmin, _vmax = contrast_stretching(data, _vmin, _vmax)
+            if centre:
+                _vmin, _vmax = centre_colormap_values(_vmin, _vmax)
```
It looks to me that we should ignore `centre` when using `overlay=True`. We should raise a warning when this option is set and ignored.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -600,7 +608,27 @@ def plot_images(images,
+    overlay : boolean, optional
+        If True, overlays the images with different colors rather than plotting
+        each image as a subplot.
+    overlay_title: boolean, optional
```
Can it be merged with the `suptitle` parameter?

ericpre reviewed (changes requested):
This is getting very nice! I have left a few minor comments. The `vmax` works well but not the `vmin` because of the transparency applied to the minimum value.

TomSlater committed:
Updates based on further suggestions

Added pixsize_factor argument.

Removed use of rcParams['image.cmap'].

Changed colormap definition for overlays.

Removed overlay_title argument.

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -488,8 +490,14 @@ def plot_images(images,
+                alphas=1.0,
+                legend_picking=True,
+                legend_loc='upper right',
+                pixsize_factor = 10,
```
Set its default value to 1 and rename to `pixel_size_factor` to avoid using abbreviation?

ericpre commented on hyperspy/drawing/utils.py:
```diff
@@ -600,7 +608,28 @@ def plot_images(images,
+        This parameter controls where the legend is placed on the figure;
+        see the pyplot.legend docstring for valid values
+    pixsize_factor : int, optional
+        Sets the size of the figure when plotting an overlay image. The higher
```
A float can also be a valid input and it would be good to specify the default value. Mention that this parameter will be ignored if `fig` if provided.

ericpre reviewed (changes requested):
A few last details in comments.
What do you want to do for the `vmin` issue? The simplest would be to ignore it and add a warning to the logger that `vmin` is ignored.

ericpre commented:
The test failures should be fixed in https://github.com/hyperspy/hyperspy/pull/2603.

TomSlater committed:
Remove vmin and update pixel_size_factor.

TomSlater commented:
> A few last details in comments.
> What do you want to do for the `vmin` issue? The simplest would be to ignore it and add a warning to the logger that `vmin` is ignored.

I think you're right. I've just ignored it. In most cases where this is used (elemental maps), it's probably not going to be an issue.

All else should be sorted now.
